helix-core/src/main/java/org/apache/helix/monitoring/mbeans/ClusterStatusMonitor.java

```diff
@@ -55,6 +55,60 @@
 import org.slf4j.LoggerFactory;
 
 public class ClusterStatusMonitor implements ClusterStatusMonitorMBean {
+  private class AsyncMissingTopStateMonitor extends Thread {
+    private final ConcurrentHashMap<String, Map<String, Long>> _missingTopStateResourceMap;
+    private long _missingTopStateDurationThreshold = Long.MAX_VALUE;;
+
+    public AsyncMissingTopStateMonitor(ConcurrentHashMap<String, Map<String, Long>> missingTopStateResourceMap) {
+      _missingTopStateResourceMap = missingTopStateResourceMap;
+    }
+
+    public void setMissingTopStateDurationThreshold(long missingTopStateDurationThreshold) {
+      _missingTopStateDurationThreshold = missingTopStateDurationThreshold;
+    }
+
+    @Override
+    public void run() {
+      try {
+        synchronized (this) {
```

Comment:
So other methods are not synchronized hence other thread (ClusterStatusMonitor) thread can still call these methods without need of acquiring thread objects's lock. (https://users.cs.fiu.edu/~weiss/cop4338_spr06/lectures/Threads/threads.pdf See slide 19). IMHO it's okay because setMissingTopStateDurationThreshold is not updating any shared resource. And reset method is like reseting the resource map which is again okay.

Reply:
Then why we want synchronized...?

Reply:
@xyuanlu because it's needed for wait() and notify() mechanism.

Reply:
You just need to lock the part where we check if condition is met. Not the whole while(true) block.